### DISCUSSION TITLE
Frigate+ model not downloaded

### DISCUSSION THREAD

**Discussion by zoechi** [blakeblackshear/frigate#14367] [Ask A Question]
I run Frigate on NixOS and just tried to configure a Plus model, but I get an error shortly after Frigate (as systemd service) is started.

```
Okt 15 16:07:01 smarthome python3.12[6503]: *************************************************************
Okt 15 16:07:01 smarthome python3.12[6503]: *************************************************************
Okt 15 16:07:01 smarthome python3.12[6503]: ***    Your config file is not valid!                     ***
Okt 15 16:07:01 smarthome python3.12[6503]: ***    Please check the docs at                           ***
Okt 15 16:07:01 smarthome python3.12[6503]: ***    https://docs.frigate.video/configuration/index     ***
Okt 15 16:07:01 smarthome python3.12[6503]: *************************************************************
Okt 15 16:07:01 smarthome python3.12[6503]: *************************************************************
Okt 15 16:07:01 smarthome python3.12[6503]: ***    Config Validation Errors                           ***
Okt 15 16:07:01 smarthome python3.12[6503]: *************************************************************
Okt 15 16:07:01 smarthome python3.12[6503]: [Errno 2] No such file or directory: '/config/model_cache/373d2276296b9a115937117a0eb568e7'
Okt 15 16:07:01 smarthome python3.12[6503]: Traceback (most recent call last):
Okt 15 16:07:01 smarthome python3.12[6503]:   File "/nix/store/6b2vhmd9v9acxglh6r0nqwhw73si7f7b-frigate-0.14.1/lib/python3.12/site-packages/frigate/app.py", line 645, in start
Okt 15 16:07:01 smarthome python3.12[6503]:     self.init_config()
Okt 15 16:07:01 smarthome python3.12[6503]:   File "/nix/store/6b2vhmd9v9acxglh6r0nqwhw73si7f7b-frigate-0.14.1/lib/python3.12/site-packages/frigate/app.py", line 140, in init_config
Okt 15 16:07:01 smarthome python3.12[6503]:     self.config = user_config.runtime_config(self.plus_api)
Okt 15 16:07:01 smarthome python3.12[6503]:                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Okt 15 16:07:01 smarthome python3.12[6503]:   File "/nix/store/6b2vhmd9v9acxglh6r0nqwhw73si7f7b-frigate-0.14.1/lib/python3.12/site-packages/frigate/config.py", line 1599, in runtime_config
Okt 15 16:07:01 smarthome python3.12[6503]:     config.model.check_and_load_plus_model(plus_api)
Okt 15 16:07:01 smarthome python3.12[6503]:   File "/nix/store/6b2vhmd9v9acxglh6r0nqwhw73si7f7b-frigate-0.14.1/lib/python3.12/site-packages/frigate/detectors/detector_config.py", line 93, in check_and_load_plus_model
Okt 15 16:07:01 smarthome python3.12[6503]:     with open(self.path, "wb") as f:
Okt 15 16:07:01 smarthome python3.12[6503]:          ^^^^^^^^^^^^^^^^^^^^^
Okt 15 16:07:01 smarthome python3.12[6503]: FileNotFoundError: [Errno 2] No such file or directory: '/config/model_cache/373d2276296b9a115937117a0eb568e7'
Okt 15 16:07:01 smarthome python3.12[6503]: *************************************************************
Okt 15 16:07:01 smarthome python3.12[6503]: ***    End Config Validation Errors                       ***
Okt 15 16:07:01 smarthome python3.12[6503]: *************************************************************
```
```
birdseye:
  enabled: true
  mode: continuous
  restream: true
cameras:
  cam_east_north:
    ffmpeg:
      inputs:
      - input_args: preset-rtsp-restream
        path: rtsp://[fd00::2:12]:8554/cam_east_north
        roles:
        - audio
      - input_args: preset-rtsp-restream
        path: rtsp://[fd00::2:12]:8554/cam_east_north_sub
        roles:
        - detect
        - record
      output_args:
        record: preset-record-generic-audio-aac
    live:
      stream_name: cam_east_north
    motion:
      contour_area: 10
      mask:
      - 0.284,0.086,0.295,0,0.434,0,0.449,0.108,0.439,0.149,0.402,0.182,0.343,0.105,0.303,0.147,0.281,0.218
  cam_garage_north:
    ffmpeg:
      inputs:
      - input_args: preset-rtsp-restream
        path: rtsp://[fd00::2:12]:8554/cam_garage_north
        roles:
        - audio
      - input_args: preset-rtsp-restream
        path: rtsp://[fd00::2:12]:8554/cam_garage_north_sub
        roles:
        - detect
        - record
      output_args:
        record: preset-record-generic-audio-aac
    live:
      stream_name: cam_garage_north
    motion:
      mask:
      - 0.025,0.334,0,0.426,0,0.692,0.07,0.697,0.12,1,0,1,0,0.285
  cam_west:
    ffmpeg:
      inputs:
      - input_args: preset-rtsp-restream
        path: rtsp://[fd00::2:12]:8554/cam_west
        roles:
        - audio
      - input_args: preset-rtsp-restream
        path: rtsp://[fd00::2:12]:8554/cam_west_sub
        roles:
        - detect
        - record
      output_args:
        record: preset-record-generic-audio-aac
    live:
      stream_name: cam_west
    motion:
      contour_area: 20
      mask:
      - 0.232,0.002,0.23,0.27,0.33,0.305,0.39,0.394,0.462,0.299,0.556,0.347,0.57,0.152,0.645,0
database:
  path: /var/lib/frigate/frigate.db
ffmpeg: {}
model:
  path: plus://373d2276296b9a115937117a0eb568e7
mqtt:
  enabled: true
  host: mqtt-ipv6.zburbach.eu
  password: '{FRIGATE_MQTT_PASSWORD}'
  user: frigate
objects:
  track:
  - person
  - car
  - dog
  - cat
  - deer
record:
  enabled: true
  events:
    post_capture: 8
    pre_capture: 5
    retain:
      default: 400
      mode: active_objects
  retain:
    days: 11
    mode: all
review:
  alerts:
    labels:
    - person
    - car
    - dog
    - deer
snapshots:
  bounding_box: true
  enabled: true
  retain:
    default: 400
  timestamp: true
timestamp_style:
  format: '%Y-&m-%d %H:%M:%S'
  position: tl
ui:
  date_style: short
  strftime_fmt: '%Y-%m-%d %H:%M'
  time_format: 24hour
  time_style: short
```  
There is an empty directory `var/cache/frigate/model_cache`

What am I missing?

**Comment by NickM-27** [comment]
something is broken with the configuration because the path is being set but the model is not there which means the model is being downloaded somewhere else. the way you are running frigate isn't an officially supported configuration so I am not aware of where that might be. running `find / -name '373d2276296b9a115937117a0eb568e7'` find figure it out

**Reply by zoechi** [reply]
This didn't find such a file. How can I download it manually?

**Reply by NickM-27** [reply]
it's a bit complicated as you'd need to get an API token. Are you running in docker or some other container, what is that config?

**Reply by zoechi** [reply]
Without Docker - Frigate is running as a SystemD service

**Reply by zoechi** [reply]
Should there be some entry in the logs about the failing model download, because I couldn't find anything. There was only the above error about the file not existing.

**Reply by NickM-27** [reply]
it will specifically try to install to /config/model_cache, perhaps it fails due to permission issues or something.

**Reply by zoechi** [reply]
Is there a way to configure where it should look for the `model_cache`.
The Systemd unit config configures `CONFIG_FILE` and `HOME` environment variables
```sh
Environment="CONFIG_FILE=/nix/store/aw1zgpbf058j5ldcd8zc0div7sg48ga0-frigate.yml"
Environment="HOME=/var/lib/frigate"
```
Perhaps I need to set SystemD `RootDirectory` to `/var/lib/frigate` or `/var/cache/frigate`.

Frigate on NixOS seems quite popular, but I couldn't find anything about configuring the Frigate+ model yet.

**Reply by NickM-27** [reply]
@blakeblackshear might have some input on how best to handle this

**Reply by blakeblackshear** [reply]
There are several files stored in the config directory, so I would suggest figuring out how to get a /config directory working properly.

**Reply by zoechi** [reply]
I have no idea how this comment should help. 
Frigate works for months. I'm obviously not the only one who doesn't know how what needs to be done because I just installed it using the Nix recipe.

Is there a way to tell Frigate where to look for the config directory like there is for the config file or is it hardcoded?

How can I download the model manually. Getting a token cant be that hard.

**Reply by NickM-27** [reply]
The problem is even if the Frigate+ model is figured out there will still be other problems running Frigate correctly with no /config directory. The config directory is hard coded and expected to be in /config due to the way Frigate is deployed. 

It sounds like the Nix recipe is incorrectly implemented. In this case it is being deployed in an entirely unsupported way and in a way that is not correctly mirroring the supported setup. The recommendation here would be to adjust the installation so it correctly mirrors a proper Frigate install

**Reply by zoechi** [reply]
Frigate worked well for several months now, so I doubt there is a bigger issue.
The problem is, that I don't know what's going on under the hood, so it's difficult to adjust my configuration. 

The path to the config file (`frigate.yaml`) is provided read-only using the `CONFIG_FILE` einvironment variable from outside.

If Frigate had tried to access `/config` directly before, I would have run into error messages or crashes. So far it seems loading the Frigate Plus model is the only thing where this directory is explicitely accessed. 
From what I found, mounting a directory just for one service is not directly supported by systemd, so I was hoping there was a better way. I try to avoid having a `/config` directory laying around in the root of my filesystem just for one file of one service.

**Reply by NickM-27** [reply]
> So far it seems loading the Frigate Plus model is the only thing where this directory is explicitely accessed.

no, that is wrong. Frigate stores many things in /config including

`.timeline` to keep track of the timeline optimizations
`.vacuum` to keep track of the db vacuum time, without this Frigate will run vacuum on every startup
`.jwt_secret` for auth
`notifications.pem` for notifications in the upcoming version
`.exports` to know if previously exported items were migrated for 0.14
`.search_stats.json` for search in the upcoming version

**Reply by zoechi** [reply]
I mounted `/var/lib/frigate/config` to `/config` for now and it's working with the Plus model.
A better solution would be appreciated, especially also one that can be added to nixpkgs without people frowning upon it.
